### PR TITLE
Signup: Pre-fill domain name with site title

### PR DESCRIPTION
**Pull request by Copons**
In the first step of the signup flow (`about`), the user can fill out their site title.
In the following step (`domain`), the user is asked to search and choose their site address.

Currently the domain search field doesn't have any initial value, even though almost always the site title and address are as similar as possible.

This PR pre-fills the domain search field with the site title the user inserted in the previous step.

## Testing Instructions

Start the signup flow from scratch (typically on `/start/about`).
Fill in the site title and proceed to next step.
The domain field should be populated with the site title, and there should already be some search results displayed.

## Notes

I had to add an `_isMounted` check to prevent firing the `onSearchChange` from `<Search>` when `<RegisterDomainStep>` wasn't fully mounted.
This happened because now, when the site title is set, we pass it as a default value to the search field, triggering that callback before its parent component is ready.

On a related note: pre-populating the `<Search>` input field throws a console warning:
> Failed prop type: You provided a `value` prop to a form field without an `onChange` handler. This will render a read-only field. If the field should be mutable use `defaultValue`. Otherwise, set either `onChange` or `readOnly`.

This is caused by the fact that `<Search>` uses `onInput` instead of `onChange` as a workaround for a [bug on IE11](https://github.com/Automattic/wp-calypso/blob/master/client/components/search/index.jsx#L367).

It might be easily fixable by replacing `value` with `defaultValue`, but despite being a one liner, there is a concrete risk of screwing up half Calypso.

### CODE REVIEW

**Comment by gwwar** [comment]
@Copons do you mind adding a gif of the updated behavior? If this passes review, let's verify with nux on changes before landing. Changes in sign up may need to be abtested.

cc @anneforbush

**Comment by fditrapani** [comment]
> Changes in sign up may need to be abtested.

We should definitely test this in an a/b test to measure the impact otherwise, how would we know if it was worth the effort?

cc @lucasartoni as we were discussing this update.

**Comment by lucasartoni** [comment]
Correct, it’s a good practice to a/b test this feature to correctly evaluate the impact. Please
Make sure there is no interference with th current testing happening on the domain search step.

**Comment by kwight** [comment]
> We should definitely test this in an a/b test to measure the impact otherwise, how would we know if it was worth the effort?

I admit this thinking feels a little dogged to me, needing AB test data to confirm what feels like a clear UX improvement. If an AB test were to fail, what would the conclusion/explanation be? Does that mean we now need tests for any field we auto-populate in signup?

I certainly find value in some AB testing, particularly for flows such as purchasing and signup that directly affect the bottom line – but I guess I find it difficult to strike a balance of "data-informed" and effectively moving signup development forward when such time and testing is required. 🙃

**Comment by kwight** [comment]
> We are currently in the middle of a complete rebooting of the domain search experience on WordPress.com. This rebooting involves a complete redesign of the search interface and a fine tuning of the suggestion engines that are under the hood.

Oh! Wow, that's great – thanks for explaining, I can see how this PR is right in the middle of all that :)

**Comment by matthusby** [comment]
> Depending on your connection, the subdomain can take a second to appear if you go back from the domain search results, update my site title, and then return to the search results.

The subdomain endpoint is a separate one from the domain search endpoint and can sometimes take longer to return, so that part is normal.

**Comment by Copons** [comment]
Thank you all for the comments. 🙇 
I wasn't aware of such a big NUX rework, and how tied NUX is with A/B tests.
I'll put this PR on hold for now to take a bit of time to read and regroup.

**Comment by Copons** [comment]
> Any edits to the domain search term are disregarded If you edit the domain name, go back to the first screen, and then return to the search result.

@fditrapani I've been thinking about this for a bit, and I'm not sure we have a mean to make it work as you suggest—which is indeed what I'd expect as a user.

This is what I understand happens:

- We start at `/start/about`, and fill the site title with `"foo"`.
  - Behind the scene, we save in local storage something like:
`{ stepName: 'about', providedDependencies: { siteTitle: 'foo' }, lastUpdated: 1111 }`
- We proceed to `/start/domains` which is now pre-filled with `"foo"` (obtained from the Redux state).
- We modify the domain query into `"foo bar"`.
  - We add to the local storage something like:
`{ stepName: 'domains', domainForm: { lastDomainSearched: 'foo bar' }, lastUpdated: 2222 }`
- Now we go back to modify some info in the `/start/about` step.
  - The first step in local storage updates regardless of the fact we actually did some changes:
`{ stepName: 'about', providedDependencies: { siteTitle: 'foo' }, lastUpdated: 3333 }`
- Onward again to `/start/domains`, which gets pre-filled again with the site title fetched from Redux state, overwriting our previous `"foo bar"` search.

So, we have multiple possibilities that can happen from the user standpoint:

- User goes back to actually change the site title, and so the overwrite is useful for their purposes.
- User goes back and changes the title, but the fact that the domain query changes unexpectedly leaves them confused.
- User goes back, doesn't change the title, and loses whatever query they were working with in the `domains` step.

But also from ours:

- If the `about` step's timestamp gets updated every time the user submits its form, how are we supposed to know if they have actually changed the site title?
- We might want to check the `siteTitle` prop on `willReceiveProps`, and only suggest it as domain query if it changed (I've yet to explore this, though).

Of course any suggestion would be welcome (there are some parts of the signup flow that are still very obscure to me 😆)

**Comment by Copons** [comment]
I've added an A/B test in 9415279.
For more info and discussion about it: p7jreA-1vS-p2 (internal).

**Comment by jsnmoon** [comment]
[Auto-fill search box with site name from NUX survey](https://trello.com/c/hj1lCPOx/9-auto-fill-search-box-with-site-name-from-nux-survey)

**Comment by jsnmoon** [comment]
Nice work, @Copons! The code looks reasonable and the `/start` flow works as expected both in development and (dockerized) production environments (1). To be safe, I will defer to @Automattic/cobalt for their domain expertise (pun). 

A more involved way to address the issues you've mentioned [here](https://github.com/Automattic/wp-calypso/pull/22376#issuecomment-366342748) would be to conditionally set the domain search input value in the [about step](https://github.com/Automattic/wp-calypso/blob/94152796e78109d15da73b48d58ca3733ec96b27/client/signup/steps/about/index.jsx#L261) using a Redux action (e.g. prefill the domain search input value only once). By doing this, we can prevent unintentionally overwriting the user's domain search query.

Footnotes:
1) I locally modified the `datestamp` value for the A/B test in `active-tests.js` to ensure that the variation assignment was being performed properly.

**Comment by fditrapani** [comment]
Sorry about the delay in responding @Copons — I've been AFK. You brought up a number of good use cases. I would go with something along the lines of @jsnmoon's suggestion:

> conditionally set the domain search input value in the about step using a Redux action (e.g. prefill the domain search input value only once). By doing this, we can prevent unintentionally overwriting the user's domain search query.

**Comment by jsnmoon** [comment]
Commandeering this PR as per @Copons request (who will be AFK for a bit).

**Comment by jsnmoon** [comment]
I've reworked this implementation by using a Redux state to keep track of the domain search input's prefill value. Doing this circumvents issues mentioned [here](https://github.com/Automattic/wp-calypso/pull/22376#issuecomment-366342748) by [preventing overwrites](https://github.com/Automattic/wp-calypso/pull/22376/files#diff-1eae7f3f2c7e120be1c7627ba034dd29R15) to the prefill value in the reducer.

The downside to this approach: This requires a lot more code than the original approach. On the bright side: At least some of the added code has test coverage.

Reviewers: What do you think?

**Comment by kwight** [comment]
@jsnmoon this doesn't seem to work for me, I've got an empty domain search after adding a site title. Are we missing initial values or something?

**Comment by kwight** [comment]
> The downside to this approach: This requires a lot more code than the original approach. On the bright side: At least some of the added code has test coverage.

Tests 🎉!

This PR /has/ gotten quite complicated and heavy; I admit I'm inclined to keep it on the simpler side that @Copons started with, but appreciate the Redux approach takes care of other issues. I think in general, I find signup code to be challenging to work with, so every approach feels heavier and more awkward than it could be. But solving that problem is bigger than anything we're trying to do here :)

**Comment by jsnmoon** [comment]
Hi @kwight, it's most likely due to the fact that the [A/B test] is defined to start on February 26th. If you change the datestamp to one from the past (ex. `20180201`) and change the allocation to `noPrefill: 0, withSiteTitle: 100,`, you'll see the expected behavior.

I share the same concern regarding code complexity. I wonder if it'll be worthwhile to keep this approach vs @Copons' original approach.

**Comment by fditrapani** [comment]
I'm also having issues getting this test to show even after updating with your instructions @jsnmoon. I noticed some weird stuff happening with a/b tests over the last month. Maybe a rebase might help?

**Comment by jsnmoon** [comment]
Ah, strange; let me investigate further.

**Comment by jsnmoon** [comment]
🤦‍♂️ I mislabeled one of the action function parameters... I added two tests to ensure this won't happen again.

**Comment by jsnmoon** [comment]
As far as I could tell after talking with @hambai, the above issue is pre-existing and most likely has to do with how [this](https://github.com/Automattic/wp-calypso/blob/master/client/lib/signup/progress-store.js) code behaves. Definitely worth investigating in a separate issue, but out-of-scope for this PR.

File: client/state/signup/steps/domains/reducer.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+import { domainStepSchema } from './schema';
+
+export default createReducer(
+	{},
+	{
+		[ SIGNUP_STEPS_DOMAIN_SEARCH_PREFILL_SET ]: ( state, action ) => {
+			const { overwrite, prefill } = action;
+
+			// Only set prefill if the value hasn't already been set.
+			if ( typeof state.prefill === 'string' && state.prefill.length > 0 && ! overwrite ) {
```
Comment (Copons) [inline client/state/signup/steps/domains/reducer.js]: Probably overkill to test if it's a string?
Maybe we can default to it in the action?

```es6
export function setDomainSearchPrefill( prefill = '', overwrite = false ) { ... }
```

**Comment by Copons** [comment]
This LGTM, except a super minor nitpick.
Thanks again @jsnmoon for taking control of this PR! 🙇 

> the above issue is pre-existing

Can confirm.
I had that behaviour before starting to work on this PR, and assumed it was working as intended.

**Comment by jsnmoon** [comment]
Hi @vindl, correct - that is an intended side-effect. We want to avoid overriding a user's input by being clever with the UI. This was the original intent of the PR as explained by @Copons [here](https://github.com/Automattic/wp-calypso/pull/22376#issuecomment-366342748).

**Comment by jsnmoon** [comment]
Could I get an explicit 👍/🚢 from someone in @Automattic/cobalt for this PR?

**Pull request merged**